packages/frontend/src/hooks/cartesianChartConfig/useCartesianChartConfig.ts

```diff
@@ -490,105 +492,23 @@ const useCartesianChartConfig = ({
-                                            {
-                                                field: pivotKey,
-                                                value: rawValue,
-                                            },
-                                        ],
-                                    },
-                                },
-                                areaStyle: defaultAreaStyle,
-                                stack,
-                            };
-                            return {
-                                ...acc,
-                                [getSeriesId(pivotSeries)]: pivotSeries,
-                            };
-                        }, {});
-
-                        return { ...sum, ...groupSeries };
-                    }, {});
-                } else {
-                    expectedSeriesMap = (dirtyLayout.yField || []).reduce<
-                        Record<string, Series>
-                    >((sum, yField) => {
-                        const series = {
-                            encode: {
-                                xRef: { field: dirtyLayout.xField },
-                                yRef: {
-                                    field: yField,
-                                },
-                            },
-                            type: defaultCartesianType,
-                            areaStyle: defaultAreaStyle,
-                            stack:
-                                isStacked || !!defaultAreaStyle
-                                    ? 'stack-all-series'
-                                    : undefined,
-                        };
-                        return { ...sum, [getSeriesId(series)]: series };
-                    }, {});
-                }
-                const existingValidSeriesMap =
-                    prev?.series?.reduce<Record<string, Series>>(
-                        (sum, series) => {
-                            if (
-                                !Object.keys(expectedSeriesMap).includes(
-                                    getSeriesId(series),
-                                )
-                            ) {
-                                return { ...sum };
-                            }
-                            return {
-                                ...sum,
-                                [getSeriesId(series)]: series,
-                            };
-                        },
-                        {},
-                    ) || {};
+                let expectedSeriesMap = getExpectedSeriesMap({
+                    defaultAreaStyle,
+                    defaultCartesianType,
+                    availableDimensions,
+                    isStacked,
+                    pivotKey,
+                    resultsData,
+                    xField: dirtyLayout.xField,
+                    yFields: dirtyLayout.yField,
+                });
+                let newSeries = mergeExistingAndExpectedSeries({
```

Comment:
Moved expected series logic to util function. 
Changed logic to add the new series in the correct order.